### ISSUE TITLE
cfstream_test fails after c-ares was enabled

### ISSUE THREAD

**Issue by rmstar** [grpc/grpc#18080]
cfstream_test fails after c-ares resolver was enabled. The test passes if we set GRPC_DNS_RESOLVER=native.

```
[ RUN      ] CFStreamTest.NetworkTransition
10.0.0.1      grpctest
dyld: lazy symbol binding failed: Symbol not found: _ares_library_init
  Referenced from: /Volumes/BuildData/tmpfs/tmp/bazel/a002d9727939acebf422134be2d181ab/execroot/com_github_grpc_grpc/bazel-out/darwin-fastbuild/bin/test/cpp/end2end/cfstream_test.runfiles/com_github_grpc_grpc/test/cpp/end2end/../../../_solib_darwin_x86_64//liblibgrpc_Uresolver_Udns_Uares.so
  Expected in: flat namespace
dyld: Symbol not found: _ares_library_init
  Referenced from: /Volumes/BuildData/tmpfs/tmp/bazel/a002d9727939acebf422134be2d181ab/execroot/com_github_grpc_grpc/bazel-out/darwin-fastbuild/bin/test/cpp/end2end/cfstream_test.runfiles/com_github_grpc_grpc/test/cpp/end2end/../../../_solib_darwin_x86_64//liblibgrpc_Uresolver_Udns_Uares.so
  Expected in: flat namespace
*******************************
Caught signal SIGABRT
  no backtrace
Abort trap: 6
```
https://source.cloud.google.com/results/invocations/f5897734-9b6a-4ef9-9466-a7ab1f101749/targets/grpc%2Fcore%2Fmaster%2Fmacos%2Fgrpc_cfstream/log

**Comment by rmstar** [comment]
CC @apolcyn

**Comment by srini100** [comment]
@rmstar, this is probably not related to CFStream itself but something to do with Bazel and cares symbols. Please confirm if this works without CFStream enabled.

**Comment by rmstar** [comment]
@srini100 yes, this issue seems to be unrelated to CFStream itself. The test fails even when CFStream is disabled.

**Comment by devjgm** [comment]
This also breaks for the https://github.com/googleapis/google-cloud-cpp project. I changed our project to use gRPC v1.19.0, then our tests on Mac OS X started failing with the same error as above. Our full error log is here:  https://source.cloud.google.com/results/invocations/9e470476-78aa-4bd2-9de9-8306722ccc82/targets/cloud-cpp%2Fgithub%2Fmacos%2Fpresubmit/log

Here's a snippet of the error:

```console
[ RUN      ] NoexTableAsyncReadRowsTest.Simple
dyld: lazy symbol binding failed: Symbol not found: _ares_library_init
  Referenced from: /Volumes/BuildData/tmpfs/tmp/bazel/958e32d54e02d3b435828e68d1d5184e/sandbox/darwin-sandbox/2133/execroot/com_github_googleapis_google_cloud_cpp/bazel-out/darwin-fastbuild/bin/google/cloud/bigtable/bigtable_client_internal_table_async_row_reader_test.runfiles/com_github_googleapis_google_cloud_cpp/google/cloud/bigtable/../../../_solib_darwin_x86_64//libexternal_Scom_Ugithub_Ugrpc_Ugrpc_Slibgrpc_Uresolver_Udns_Uares.so
  Expected in: flat namespace
dyld: Symbol not found: _ares_library_init
  Referenced from: /Volumes/BuildData/tmpfs/tmp/bazel/958e32d54e02d3b435828e68d1d5184e/sandbox/darwin-sandbox/2133/execroot/com_github_googleapis_google_cloud_cpp/bazel-out/darwin-fastbuild/bin/google/cloud/bigtable/bigtable_client_internal_table_async_row_reader_test.runfiles/com_github_googleapis_google_cloud_cpp/google/cloud/bigtable/../../../_solib_darwin_x86_64//libexternal_Scom_Ugithub_Ugrpc_Ugrpc_Slibgrpc_Uresolver_Udns_Uares.so
  Expected in: flat namespace
Abort trap: 6
================================================================================
```

**Comment by srini100** [comment]
@devjgm, sorry we missed this issue in 1.19.0. Are you ok with pinning to 1.17.2 for now? I would like to avoid a patch release just for this issue.

**Comment by apolcyn** [comment]
Is there a way to test on master branch? The fix (https://github.com/grpc/grpc/pull/18201) is now on master branch

**Comment by devjgm** [comment]
The problem is that I can't currently build grpc with the current bazel due to an old protobuf dependency. Specifically, I need a version of grpc with https://github.com/grpc/grpc/commit/7a814e2597fbda08e06034717324d8e5b423b93a#diff-5952172e3b803ed640371cda4cddf13d I believe 1.19.0 has this, which is why I need to upgrade.

Regarding avoiding a patch release: isn't the 1.19.0 build broken on some OS X builds? If so, shouldn't that be fixed? Thanks.

**Comment by srini100** [comment]
@devjgm, is there a way for you to confirm the fix in #18201. We'll target to release a patch next week. This affects bazel but not make builds on MacOS.

**Comment by devjgm** [comment]
Sure. I tried building against grpc @ commit 046e3e4ab5cdb867bda7a5c40fd9b0ac84ad0ff8. The tarball bazel fetched was https://github.com/grpc/grpc/archive/046e3e4ab5cdb867bda7a5c40fd9b0ac84ad0ff8.tar.gz

When I use this grpc snapshot, I get the following error:

```shell
DEBUG: Rule 'com_google_absl' modified arguments {"sha256": "5fe2a3a8f8378e81d4d3db6541f48030e04233ccd2d6c7e9d981ed577b314ae8"}
ERROR: /home/jgm/github/devjgm/google-cloud-cpp/google/cloud/bigtable/examples/opencensus/BUILD:19:1: error loading package '@io_opencensus_cpp//opencensus/exporters/trace/stdout': in /home/jgm/.cache/bazel/_bazel_jgm/bd671f688e1f702fab062681330ab451/external/io_opencensus_cpp/opencensus/copts.bzl: Unable to load file '@com_google_absl//absl:copts.bzl': file doesn't exist and referenced by '//google/cloud/bigtable/examples/opencensus:bigtable_opencensus'
ERROR: Analysis of target '//google/cloud/bigtable/examples/opencensus:bigtable_opencensus' failed; build aborted: error loading package '@io_opencensus_cpp//opencensus/exporters/trace/stdout': in /home/jgm/.cache/bazel/_bazel_jgm/bd671f688e1f702fab062681330ab451/external/io_opencensus_cpp/opencensus/copts.bzl: Unable to load file '@com_google_absl//absl:copts.bzl': file doesn't exist
INFO: Elapsed time: 0.286s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (0 packages loaded, 0 targets configured)
    currently loading: @io_opencensus_cpp//opencensus/exporters/trace/stdout
```

So it looks like I'm picking up so other commits between the 1.19.0 release and the commit from #18201 that is changing something that may be causing that breakage.

Pointers?

**Comment by rmstar** [comment]
@devjgm the error you pointed out seems to be related to a package dependency on abseil. I don't think it's caused by the c-ares linking issue.

I noticed that commit 046e3e4 and v1.19.0 pull in different abseil-cpp tarballs. Not sure if that's causing the breakage.

[v.19.0
](https://github.com/grpc/grpc/blob/v1.19.0/bazel/grpc_deps.bzl#L170-L175)
```
    if "com_google_absl" not in native.existing_rules():
        http_archive(
            name = "com_google_absl",
            strip_prefix = "abseil-cpp-cd95e71df6eaf8f2a282b1da556c2cf1c9b09207",
            url = "https://github.com/abseil/abseil-cpp/archive/cd95e71df6eaf8f2a282b1da556c2cf1c9b09207.tar.gz",
        )

```
[commit 046e3e4
](https://github.com/grpc/grpc/blob/046e3e4ab5cdb867bda7a5c40fd9b0ac84ad0ff8/bazel/grpc_deps.bzl#L170-L175)

```
    if "com_google_absl" not in native.existing_rules():
        http_archive(
            name = "com_google_absl",
            strip_prefix = "abseil-cpp-308ce31528a7edfa39f5f6d36142278a0ae1bf45",
            url = "https://github.com/abseil/abseil-cpp/archive/308ce31528a7edfa39f5f6d36142278a0ae1bf45.tar.gz",
        )

```

**Comment by srini100** [comment]
Reopening since there is active discussion going on.

**Comment by remyabel** [comment]
abseil moved the location of `configure_copts.bzl` to `copts.bzl`. The version of io_opencensus_cpp that google-cloud-cpp loads is probably trying to find the old location. I will open a PR for google-cloud-cpp updating the dependency and see if it fixes the issue.

**Comment by srini100** [comment]
@devjgm, @remyabel, are you able to verify the original fix in your local build? We want to do an RC for 1.19.1 patch this week.

**Comment by remyabel** [comment]
Yes the original fix works. Thanks @srini100.

**Comment by srini100** [comment]
There is some delay in rolling this patch out. Python bazel on macos seems to have similar issue. See https://github.com/grpc/grpc/issues/18267. Would like to get that resolved too.

**Comment by apolcyn** [comment]
Note the error in #18267 is `Symbol not found: __Z31grpc_set_default_iomgr_platformv`

That might be a similar issue to the one fixed by #18201, but it's definitely unrelated to c-ares

**Comment by srini100** [comment]
@devjgm, @remyabel, I have tagged the release [v1.19.1]( https://github.com/grpc/grpc/releases/tag/v1.19.1). Do you need any pkgs to be released or source is sufficient?
